Tap to Place Documentation (#7412)
microsoft/MixedRealityToolkit-Unity: Documentation/README_TapToPlace.md

# Tap to Place

![TapToPlace](Images/Solver/TapToPlace/TapToPlaceIntroGif.gif)

> Tap to Place is a far interaction component that is used to place a game object on surface. This component is useful for placing objects on a spatial mesh. Tap to Place uses a combination of two clicks and head movement to place an object. A click to start the placement, head movement to change the position of the object and a click to place.

Reviewer: ```suggestion
Tap to Place is a far interaction component that is used to place a game object on surface. This component is useful for placing objects on a spatial mesh. Tap to Place uses a combination of two clicks and head movement to place an object. A click to start the placement, head movement to control the position of the object and a click to place the object in the scene.
```

## Using Tap to Place

1. Set up the scene
    - Create a new unity scene

>     - Add MRTK to the scene by navigating to the Mixed Reality Toolkit menu > Add to Scene and Configure

Reviewer: ```suggestion
    - Add MRTK to the scene by navigating to the **Mixed Reality Toolkit** > **Add to Scene and Configure**
```

Author: Added

>     - Add a cube to the scene and change the scale to 0.2 and change the position to (0, 0, 0.7).

Reviewer: This should not be necessary. Solvers should be independent of MRTK

>     - Add a cube to the scene and change the scale to 0.2 and change the position to (0, 0, 0.7).

Reviewer: Add IMPORTANT note that TapToPlace requires MRTK input system

Reviewer: see section below for controlling TapToPlace without clicks driven by the MRTK input system

Author: Added

> 1. Attach [Tap to Place](https://microsoft.github.io/MixedRealityToolkit-Unity/api/Microsoft.MixedReality.Toolkit.Utilities.Solvers.TapToPlace.html) to a game object with a collider

Reviewer: this should be an xref

Reviewer: xref:Microsoft.MixedReality.Toolkit.Utilities.Solvers.SolverHandler
replace with namespace of taptoplace

Author: added

    ![TapToPlaceInspector](Images/Solver/TapToPlace/TapToPlaceInspector2.png)
    
    - When the Tap to Place component is added, a Solver Handler will also be attached. Tap to Place derives from the [Solver](README_Solver.md) class which requires a Solver Handler. The position of a Tap to Place object is calculated relative to the `TrackedTargetType` within the Solver Handler. By default the Head is the `TrackedTargetType`, i.e. when the head moves, the object follows if it is selected.  The `TrackedTargetType` can also be set to Controller Ray which has the object follow the controller. The first group of properties in the Tap to Place inspector are the [Common Solver Properties](README_Solver.md#common-solver-properties).  

>     > [!NOTE]

Reviewer: make this important. Also explain why it has to be standalone. Solvers are generally designed to be chained. But I see this one controls SolverHandler.UpdateSolvers

Author: added

    > Tap to Place is a stand alone Solver and cannot be chained with other solvers. 
    - Tap to Place Properties:

>         - `Auto Start`: If true, the game object to place will start out selected.  The object will immediately start following the TrackedTargetType (Head or Controller Ray). This value must be modified before Start() is invoked in order to have any effect.

Reviewer: ```suggestion
        - `Auto Start`: If true, the Tap to Place solver will start out controlling the position of the gameobject to be placed.  The object will immediately start following the TrackedTargetType (Head or Controller Ray). This value must be modified before Start() is invoked in order to have any effect.
```

Author: Added

        - `Magnetic Surfaces`: An array of LayerMasks to execute from highest to lowest priority. First layer mask to provide a raycast hit will be used for position calculations.
        - `Debug Enabled`: If true and in the Unity Editor, the normal of the raycast hit will be drawn in yellow. Debug enabled is useful when `RotateAccordingToSurface` is true because it draws the normal of the surface hit which visually explains why the object is set at its current orientation.
        - `On Placing Started`: This event is triggered once when the game object to place is selected.
        - `On Placing Stopped`: This event is triggered once when the game object to place is unselected, placed.

1.  Testing Tap to Place behavior in editor
    - Press play and hold the space bar to show an input simulation hand.
    - Move the hand until the cube is in focus and simulate a click with the input simulation hand by clicking with the left mouse.
        - If colliders are not present in the scene, the object will follow the `TrackedTargetType` at the defined `Default Placement Distance`.

>     - The object will follow the movement of the `TrackedTargetType` after selection. To simulate head movement in editor, press the WASD keys. Change head rotation by clicking and holding the right mouse.

Reviewer: I am tempted to say remove the WASD keys/controlling head movement. Maybe just reference the input simulation page. If things change or the user has changed their project, this just seems another area that would have to be updated and may be invalidated.

Author: I think I am going to keep it in there for now since this documentation is in the form of a tutorial and I like having all the information on one page.

    - To stop placing the object, click again.  The object does not need to be in focus for the stop placement click. Focus is only required for the initial click that starts the placement process.

    `TrackedTargetType`: Head (Default) |  `TrackedTargetType`: Controller Ray 
    :-------------------------:|:-------------------------:
    ![TapToPlaceInputSimulationHead](Images/Solver/TapToPlace/TapToPlaceInputSimulationHead.gif)  |  ![TapToPlaceInputSimulationControllerRay](Images/Solver/TapToPlace/TapToPlaceInputSimulationControllerRay.gif)

## Tap to Place Code Configurability 

> Tap to Place object selection timing can also be controlled via `StartPlacement()` and `StopPlacement()` instead of requiring a click event. This capability is useful for writing tests and provides an alternative method to place an object in editor without using the input simulation. 

Reviewer: ```suggestion
Tap to Place object selection timing can also be controlled via `StartPlacement()` and `StopPlacement()` instead of requiring a click event. This capability is useful for writing tests and provides an alternative method to place an object in editor without using the MRTK Input system. 
```

Author: Added

1. Create an empty game object
1. Create and attach the following example script to the empty game object

    ```
    using UnityEngine;
    using Microsoft.MixedReality.Toolkit.Utilities.Solvers;

    public class TapToPlaceInputExample : MonoBehaviour
    {

>         GameObject cube;

Reviewer: ```suggestion
        public GameObject target;
```

>         GameObject cube;

Reviewer: ```suggestion
        private GameObject cube;
```

Author: Added

>         TapToPlace tapToPlace;

Reviewer: ```suggestion
        public TapToPlace tapToPlace;
```

>         TapToPlace tapToPlace;

Reviewer: ```suggestion
        private TapToPlace tapToPlace;
```

Author: Added

        void Start()
        {
            cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
            cube.transform.localScale = Vector3.one * 0.2f;
            cube.transform.position = Vector3.forward * 0.7f;

            tapToPlace = cube.AddComponent<TapToPlace>();
        }

        void Update()
        {
            if (Input.GetKeyDown(KeyCode.U))
            {
                tapToPlace.StartPlacement();
            }
            if (Input.GetKeyDown(KeyCode.I))
            {
                tapToPlace.StopPlacement();
            }
        }
    }
    ```

> 1. In play mode, press the U key to start placing the cube

Reviewer: ```suggestion
1. In play mode, press the *U key* to start placing the cube
```

Author: Added

> 1. Press the I key to stop the placement

Reviewer: ```suggestion
1. Press the *I key* to stop the placement
```

Author: added

## Tap to Place Example Scene

The Tap to Place example scene consists of 4 placeable objects, each with a different configuration. The example scene contains walls to show the surface placement behavior that are disabled by default in the hierarchy. The example scene can be found in the Microsoft.MixedReality.Toolkit.Unity.Examples unity package found on the [Release Page](https://github.com/Microsoft/MixedRealityToolkit-Unity/releases). The scene location is: *MixedRealityToolkit.Examples/Demos/Solvers/Scenes/TapToPlaceExample.unity*

![TapToPlaceExampleScene](Images/Solver/TapToPlace/TapToPlaceExampleScene.gif)

## See also

> * [Solvers](README_Solver.md)

Reviewer: I would also add spatial awareness

Reviewer: and maybe input simulation

Author: Added
